bug(lowering): non-deterministic heavy-module emit corruption recurs on macOS-arm64 CI (post-#892)

## Goal

Eliminate the residual non-deterministic codegen/emit corruption that still intermittently aborts macOS-arm64 CI test shards with `Error 134` (SIGABRT), despite #892 (`residual non-deterministic heap corruption in runtime-helper declare emission`) having been closed as completed on 2026-06-24. The CI test-retry harness still cites #892 by name when it auto-retries, and the flake it describes is still firing — so either #892's fix was incomplete or a sibling surface regressed.

## Context

This is a **fresh recurrence of the #892 / #740 class** of bug: a latent UAF/aliasing in the heavy-module emit path whose symptom is *non-deterministic* — the failing test set changes run-to-run because it is heap-layout/timing dependent, not a logic error. #892 was closed (completed) by `mcereal` on 2026-06-24, but the same non-determinism is still present.

### Evidence (PR #1911, CI run `28715601852`, run #1802, commit `1aaab54`)

The **macOS-arm64 `e2e-d`** shard (job `85157231295`) failed, and — the tell-tale signature — a **different set of tests failed on each of the two auto-retry attempts**, both ending in `Error 134`:

- **Attempt 1** — summary `passed:204, failed:1, skipped:0`; the single failure was:
  ```
  array_reduce_closure_test.sfn:46  ("array reduce: capturing reducer folds correctly (14, not the initial 0)")
    assertion failed @ :48:5
  make: *** [test-shard] Error 134
  [ci] tests failed on attempt 1; retrying once ... non-deterministic codegen corruption tracked in #892
  ```
- **Attempt 2** — summary `passed:201, failed:2, skipped:2`; a *different* pair failed:
  ```
  linker_selection_test.sfn:79  ("linker: auto-detect surfaces [link] on stderr (not stdout) and runs")
    assertion failed @ :92:5   (its two sibling SAILFIN_LINKER= tests then reported `skip`)
  runtime_array_test.sfn:164    ("runtime-array: heavy-module emit is deterministic across runs")
    assertion failed @ :172:9
  make: *** [test-shard] Error 134
  ```
- A **manual re-run of the failed jobs** (`rerun-failed-jobs` on the same run) then passed cleanly — confirming the failures were non-deterministic, not a real regression in any of those tests.

The three tests that failed across the two attempts (`array_reduce_closure`, `linker_selection`, `runtime_array`) are **mutually unrelated**; the only common factor is heavy-module emit under the parallel `int-e2e-caps` pool on macOS-arm64. Notably, `runtime_array_test.sfn`'s failing case is literally titled *"heavy-module emit is deterministic across runs"* — a direct hit on the property #892 was meant to guarantee.

The change under test in #1911 cannot be the cause: it adds only a guarded `print.err` in the link path that is byte-identical output when `SAILFIN_TRACE_LINK` is unset (the #1112 zero-behaviour gate), and the flag is unset for every one of these tests.

### Relationship to prior issues
- **#892** (closed/completed) — residual of **#740** (closed by #741 / `36e5cdc`). #892's specific repro (`declare i1 sailfin_runtime_is_void` `@`-byte clobber on linux-x86_64) was fixed, but the underlying heap-layout-sensitive UAF/aliasing class in the emit path appears to survive on macOS-arm64.
- Candidate root-cause surfaces are unchanged from #740/#892: `extract_all_call_targets`, `append_unique_effect`, `merge_effect_lists` (`compiler/src/llvm/effects.sfn`) — substring views/array-of-string-pointer pushes that alias a later-mutated buffer; and the declare-emission consumer `collect_runtime_helper_targets_from_lines` (`compiler/src/llvm/lowering/lowering_helpers.sfn`).

## Scope

**In:**
- Reproduce the non-determinism under macOS-arm64 load (parallel pool), then root-cause the residual UAF/aliasing in the heavy-module emit path (the #740/#892 surfaces above).
- Fix so that heap-layout shifts cannot corrupt unrelated emitted output (SIGABRT / non-byte-identical emit).
- A regression that reliably reproduces under load and fails *before* the fix (extend the existing `runtime_array_test.sfn` "heavy-module emit is deterministic across runs" sweep, and/or a descriptor-table emit-sweep harness), asserting byte-identical emit across N sweeps on **both** linux-x86_64 and macos-arm64.

**Out:**
- The #1908/#1911 `SAILFIN_TRACE_LINK` work — it merely happened to be the PR that surfaced the flake; do not revert it.
- Re-disabling helper collection (`skip_helper_collection = true`) — re-introduces the problem #740 unblocked.
- The per-shard auto-retry harness — keep it; it is the mitigation, not the fix. (Consider updating its inline `#892` reference to point here once this issue is triaged.)

## Acceptance Criteria
- [ ] A regression test reliably reproduces the corruption (SIGABRT / non-byte-identical emit) under parallel-pool load before the fix.
- [ ] After the fix, that test passes byte-identically across N sweeps under `make check` load on both linux-x86_64 and macos-arm64.
- [ ] The macOS-arm64 e2e shards run without `Error 134` flakes across repeated CI runs (spot-check several).
- [ ] `make compile && make check` clean end-to-end.

## Files Affected (advisory)
- `compiler/src/llvm/effects.sfn` (`extract_all_call_targets`, `append_unique_effect`, `merge_effect_lists`)
- `compiler/src/llvm/lowering/lowering_helpers.sfn` (`collect_runtime_helper_targets_from_lines`, declare loop)
- Possibly `runtime/prelude.sfn` / `runtime/sfn/string.sfn` (`substring` view-vs-copy on this path)
- A regression test under `compiler/tests/` (extend `compiler/tests/e2e/runtime_array_test.sfn`)

## Verification
```bash
make compile && make check
# Plus the new reliable repro test, run under macOS-arm64 parallel-pool load.
```

## Size
M — deep, non-deterministic memory diagnosis; recommend the `seed-stabilizer` agent (Opus). Same class and surfaces as #740 / #892.

## Type
Bug

## Notes on seed integrity
Non-deterministic emit corruption is a seed-integrity risk (it can silently poison an emitted `.sfn-asm`/IR that then becomes a seed), which is why #892 carried `seed-blocker`. Grooming may want to restore that label here.

## Blocked by
None.

## Required in pinned seed
None — fix is in `compiler/src/` self-host logic.

--- a/compiler/src/backend.sfn
+++ b/compiler/src/backend.sfn
@@ -7,7 +7,10 @@
 // clang path is the sole implementation (`LlvmTextBackend`); it
 // reproduces every existing clang invocation byte-for-byte (zero
 // behaviour change — `--check-determinism` must still pass with no
-// rebaseline).
+// rebaseline), with the one deliberate exception that on Darwin the link
+// now self-supplies an explicit `-isysroot`/`-L<sdk>/usr/lib` (see the
+// `_macos_sdk_link_flags` note below, #1914); the Linux invocation is
+// unchanged and the addition is memoized/deterministic per machine.
 //
 // Later codegen/link work plugs in here instead of re-hardcoding LLVM
 // assumptions across the driver: the LLVM C-API binding (#347) lands as
@@ -36,9 +39,13 @@
 // importing this module (and emitting the `LlvmTextBackend` vtable). See
 // that file's header for the seed-lowering reason (#1730).
 import { _llvm_text_assemble_argv } from "./build/clang_argv";
+// `_shell_read_cmd` runs a command and trims its stdout — used by the
+// macOS SDK probe (`uname -s`, `xcrun --show-sdk-path`).
+import { _shell_read_cmd } from "./build/fs";
 // `_env_flag` gives `SAILFIN_TRACE_LINK` the same toggle semantics ("",
-// "0", "false" are off) as the other `SAILFIN_TRACE_*` knobs.
-import { _env_flag } from "./build_flags";
+// "0", "false" are off) as the other `SAILFIN_TRACE_*` knobs;
+// `_get_env_cmd` reads a single env var (the macOS SDK probe below).
+import { _env_flag, _get_env_cmd } from "./build_flags";
 
 // Reserved opaque object-file handle for later backend stages (#347 /
 // #1640 produce these in-process). Stage 0's `LlvmTextBackend` returns
@@ -84,10 +91,71 @@ interface Backend {
     fn link(self, plan: LinkPlan) -> int ![io];
 }
 
+// macOS SDK link flags (#1914). The driver invokes `clang` by bare name
+// and, on Darwin, has historically leaned on the *ambient* environment
+// (`SDKROOT` / `LIBRARY_PATH`) for the linker to locate the platform
+// libraries (`libSystem`, `libm`, `libpthread`). That is fragile: a
+// non-Apple `clang` first on `PATH` (a common Homebrew-LLVM setup) with
+// those vars unset fails the link outright (`ld: library 'System'/'m' not
+// found`), and a subprocess build that inherits a stripped env — e.g. the
+// parallel test pool's per-child env, which forwards only PATH/HOME/TMPDIR
+// — reintroduces the same gap. That gap surfaced as non-deterministic
+// macOS-arm64 e2e failures because whether a nested build's link resolved
+// depended on load-timed SDK re-detection rather than on the source.
+// Passing an explicitly-probed `-isysroot`/`-L<sdk>/usr/lib` makes the
+// link self-sufficient regardless of which clang is on PATH or whether the
+// SDK env is present. Non-Darwin returns [] (no-op); additive only when a
+// real SDK is found, so a working Apple-clang link is byte-unchanged.
+//
+// Probed once and cached (a link may recur per test binary in one process;
+// caching also avoids re-running `xcrun` under load, the very race this
+// closes). Only scalars are memoized — matching the `errno_locator_symbol`
+// lazy-init contract in `llvm/lowering_debug_state.sfn`.
+let mut _macos_sdk_probe_done: boolean = false;
+let mut _macos_sdk_root: string = "";
+
+fn _macos_sdk_link_flags() -> string[] ![io] {
+    if !_macos_sdk_probe_done {
+        _macos_sdk_root = _probe_macos_sdk_root();
+        _macos_sdk_probe_done = true;
+    }
+    if _macos_sdk_root.length == 0 { return []; }
+    return ["-isysroot", _macos_sdk_root, "-L" + _macos_sdk_root + "/usr/lib"];
+}
+
+// Resolve the macOS SDK root, or "" when not on Darwin / none found.
+// Order: `SAILFIN_TARGET_OS` (cross-compile override) then `uname -s` to
+// gate on Darwin; then `SDKROOT` (no subprocess, most stable when set),
+// `xcrun --show-sdk-path`, and the Command Line Tools default. Each SDK
+// candidate must exist on disk before it is trusted, so a stale env var
+// cannot force a bad `-isysroot`.
+fn _probe_macos_sdk_root() -> string ![io] {
+    let mut os: string = _get_env_cmd("SAILFIN_TARGET_OS");
+    if os.length == 0 { os = _shell_read_cmd("uname -s"); }
+    if os != "Darwin" { return ""; }
+    let env_sdk = _get_env_cmd("SDKROOT");
+    if env_sdk.length > 0 {
+        if fs.exists(env_sdk) { return env_sdk; }
+    }
+    let xcrun_sdk = _shell_read_cmd("xcrun --show-sdk-path 2>/dev/null");
+    if xcrun_sdk.length > 0 {
+        if fs.exists(xcrun_sdk) { return xcrun_sdk; }
+    }
+    let clt = "/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk";
+    if fs.exists(clt) { return clt; }
+    return "";
+}
+
 // Build the clang link argv for the `sfn test` layout:
-// `clang -O0 -Wno-override-module -o OUT <test.ll> <dep.ll…> <runtime.o…> <libs…>`.
-fn _llvm_text_link_test_argv(plan: LinkPlan) -> string[] {
+// `clang -O0 -Wno-override-module <sdk_flags…> -o OUT <test.ll> <dep.ll…> <runtime.o…> <libs…>`.
+fn _llvm_text_link_test_argv(plan: LinkPlan, sdk_flags: string[]) -> string[] {
     let mut args: string[] = ["clang", plan.opt_flag, "-Wno-override-module"];
+    let mut si: int = 0;
+    loop {
+        if si >= sdk_flags.length { break; }
+        args.push(sdk_flags[si]);
+        si += 1;
+    }
     args.push("-o");
     args.push(plan.out_path);
     let mut ii: int = 0;
@@ -112,9 +180,15 @@ fn _llvm_text_link_test_argv(plan: LinkPlan) -> string[] {
 }
 
 // Build the clang link argv for the program/build layout:
-// `clang <opt> -Wno-override-module <link_flags…> <runtime.o…> <ir.ll…> -o OUT <libs…>`.
-fn _llvm_text_link_program_argv(plan: LinkPlan) -> string[] {
+// `clang <opt> -Wno-override-module <sdk_flags…> <link_flags…> <runtime.o…> <ir.ll…> -o OUT <libs…>`.
+fn _llvm_text_link_program_argv(plan: LinkPlan, sdk_flags: string[]) -> string[] {
     let mut args: string[] = ["clang", plan.opt_flag, "-Wno-override-module"];
+    let mut si: int = 0;
+    loop {
+        if si >= sdk_flags.length { break; }
+        args.push(sdk_flags[si]);
+        si += 1;
+    }
     let mut fi: int = 0;
     loop {
         if fi >= plan.link_flags.length { break; }
@@ -183,12 +257,13 @@ struct LlvmTextBackend {
     }
 
     fn link(self, plan: LinkPlan) -> int ![io] {
+        let sdk_flags = _macos_sdk_link_flags();
         if plan.test_mode {
-            let test_argv = _llvm_text_link_test_argv(plan);
+            let test_argv = _llvm_text_link_test_argv(plan, sdk_flags);
             _llvm_text_trace_link(test_argv);
             return process.run(test_argv);
         }
-        let program_argv = _llvm_text_link_program_argv(plan);
+        let program_argv = _llvm_text_link_program_argv(plan, sdk_flags);
         _llvm_text_trace_link(program_argv);
         return process.run(program_argv);
     }

--- a/compiler/tests/e2e/macos_sdk_link_flags_test.sfn
+++ b/compiler/tests/e2e/macos_sdk_link_flags_test.sfn
@@ -33,8 +33,14 @@ fn _scratch_dir() -> string ![io] {
     return dir;
 }
 
+// Forward `PATH` so `uname` resolves: `run_capture`'s empty env is the
+// *empty* environment (no PATH), under which the `uname` spawn could fail
+// and mis-report the host as non-Darwin (Copilot review, PR #1929).
 fn _is_darwin() -> boolean ![io] {
-    let _exit = process.run_capture(["uname", "-s"], []);
+    let mut e: string[] = [];
+    let path = env.get("PATH");
+    if path.length > 0 { e.push("PATH=" + path); }
+    let _exit = process.run_capture(["uname", "-s"], e);
     let out = process.capture_take_stdout();
     let _err = process.capture_take_stderr();
     return find(out, "Darwin", 0) >= 0;

--- a/compiler/tests/e2e/macos_sdk_link_flags_test.sfn
+++ b/compiler/tests/e2e/macos_sdk_link_flags_test.sfn
@@ -1,0 +1,95 @@
+// Regression e2e for #1914: on macOS the build driver must pass an
+// explicitly-probed `-isysroot <sdk>` / `-L<sdk>/usr/lib` to `clang` so the
+// link resolves the platform libraries (`libSystem`, `libm`, `libpthread`)
+// on its own — independent of the ambient SDK environment (`SDKROOT` /
+// `LIBRARY_PATH`) or which `clang` is first on `PATH`. Before the fix the
+// nested link leaned on that ambient env; a subprocess build that inherited
+// a stripped env (the parallel test pool forwards only PATH/HOME/TMPDIR) or
+// a non-Apple `clang` on `PATH` then failed the link (`ld: library
+// 'System'/'m' not found`), which surfaced as non-deterministic macOS-arm64
+// e2e failures because success depended on load-timed SDK re-detection.
+//
+// The discriminator is deterministic on Darwin: the link argv carries
+// `-isysroot` after the fix and not before, and a `![io]` runtime-linking
+// program built under a deliberately stripped env must succeed. On
+// non-Darwin the flags are a no-op, so assert they are absent — guarding
+// against accidentally emitting a macOS `-isysroot` on Linux.
+//
+// Matchers note (#842): `sfn/test`'s `expect_*` matchers are `![pure]` and
+// cannot be called from an `![io]` test; assert on captured text directly.
+import { contains, find } from "sfn/strings";
+
+fn _sfn_bin() -> string ![io] {
+    let configured = env.get("SAILFIN_BIN");
+    if configured.length > 0 { return configured; }
+    return "build/native/sailfin";
+}
+
+fn _scratch_dir() -> string ![io] {
+    let mut dir = env.get("SAILFIN_TEST_SCRATCH");
+    if dir.length == 0 { dir = "/tmp"; }
+    dir = dir + "/sfn-macos-sdk-link";
+    fs.createDirectory(dir, true);
+    return dir;
+}
+
+fn _is_darwin() -> boolean ![io] {
+    let _exit = process.run_capture(["uname", "-s"], []);
+    let out = process.capture_take_stdout();
+    let _err = process.capture_take_stderr();
+    return find(out, "Darwin", 0) >= 0;
+}
+
+// Deliberately minimal: forward only PATH (clang/ld discovery), HOME,
+// TMPDIR, and SAILFIN_TEST_SCRATCH (per-invocation build-cache isolation).
+// `SDKROOT` / `LIBRARY_PATH` are intentionally NOT forwarded — the whole
+// point is that the compiler supplies the SDK itself. `SAILFIN_TRACE_LINK`
+// echoes the resolved clang link argv to stderr.
+fn _stripped_traced_env() -> string[] ![io] {
+    let mut e: string[] = [];
+    let path = env.get("PATH");
+    if path.length > 0 { e.push("PATH=" + path); }
+    let home = env.get("HOME");
+    if home.length > 0 { e.push("HOME=" + home); }
+    let tmpdir = env.get("TMPDIR");
+    if tmpdir.length > 0 { e.push("TMPDIR=" + tmpdir); }
+    let scratch = env.get("SAILFIN_TEST_SCRATCH");
+    if scratch.length > 0 { e.push("SAILFIN_TEST_SCRATCH=" + scratch); }
+    e.push("SAILFIN_TRACE_LINK=1");
+    return e;
+}
+
+// A program that pulls the runtime link libs (`-lm`/`-lpthread`): `parallel`
+// links the concurrency runtime, exercising the exact link that failed.
+fn _write_fixture(dir: string) -> string ![io] {
+    let path = dir + "/prog.sfn";
+    fs.writeFile(path, "fn _w() -> int { return 7; }\n" + "fn main() ![io] {\n"
+        + "    let _h = parallel [\n"
+        + "        fn() -> int { return _w(); },\n"
+        + "    ];\n"
+        + "    print(\"sdk-link-ok\");\n"
+        + "}\n");
+    return path;
+}
+
+test "macos-sdk-link: build carries explicit -isysroot and links under a stripped env (#1914)" ![io] {
+    let dir = _scratch_dir();
+    let main_path = _write_fixture(dir);
+    let bin = dir + "/prog.bin";
+    let exit = process.run_capture([_sfn_bin(), "build", "-o", bin, main_path], _stripped_traced_env());
+    let _out = process.capture_take_stdout();
+    let err = process.capture_take_stderr();
+
+    if _is_darwin() {
+        // The fix: the link argv self-supplies the SDK.
+        assert contains(err, "-isysroot");
+        assert contains(err, "/usr/lib");
+        // Behavioural proof: the runtime-linking build succeeds without any
+        // ambient SDK env — the regression was this link failing.
+        assert exit == 0;
+    } else {
+        // No-op on non-Darwin: a Linux link must not carry a macOS SDK root.
+        assert ! contains(err, "-isysroot");
+        assert exit == 0;
+    }
+}

--- a/compiler/tests/unit/backend_link_libs_parity_test.sfn
+++ b/compiler/tests/unit/backend_link_libs_parity_test.sfn
@@ -51,7 +51,8 @@ fn _representative_plan() -> LinkPlan ![pure] {
 
 test "backend: test-layout builder reproduces every plan.libs entry" ![pure] {
     let plan = _representative_plan();
-    let argv = _llvm_text_link_test_argv(plan);
+    let no_sdk: string[] = [];
+    let argv = _llvm_text_link_test_argv(plan, no_sdk);
     let mut i: int = 0;
     loop {
         if i >= plan.libs.length { break; }
@@ -62,7 +63,8 @@ test "backend: test-layout builder reproduces every plan.libs entry" ![pure] {
 
 test "backend: program-layout builder reproduces every plan.libs entry" ![pure] {
     let plan = _representative_plan();
-    let argv = _llvm_text_link_program_argv(plan);
+    let no_sdk: string[] = [];
+    let argv = _llvm_text_link_program_argv(plan, no_sdk);
     let mut i: int = 0;
     loop {
         if i >= plan.libs.length { break; }
@@ -76,8 +78,9 @@ test "backend: program-layout builder reproduces every plan.libs entry" ![pure] 
 // above genuinely fail if a builder ever drops an entry.
 test "backend: parity check is meaningful (absent lib is not found)" ![pure] {
     let plan = _representative_plan();
-    let test_argv = _llvm_text_link_test_argv(plan);
-    let program_argv = _llvm_text_link_program_argv(plan);
+    let no_sdk: string[] = [];
+    let test_argv = _llvm_text_link_test_argv(plan, no_sdk);
+    let program_argv = _llvm_text_link_program_argv(plan, no_sdk);
     assert expect_eq_bool(_argv_contains(test_argv, "-lnot-a-real-lib"), false).ok;
     assert expect_eq_bool(_argv_contains(program_argv, "-lnot-a-real-lib"), false).ok;
 }

